lib/trino-parquet/src/main/java/io/trino/parquet/reader/PrimitiveColumnReader.java

```diff
@@ -178,42 +191,126 @@ public ColumnChunk readPrimitive(Field field)
+            processValues(valuesToRead, () -> {
+                readValue(blockBuilder, type);
+                definitionLevels.add(definitionLevel);
+                repetitionLevels.add(repetitionLevel);
+            });
+        }
+        else {
+            processValuesSync(valuesToRead, () -> {
+                readValue(blockBuilder, type);
+                definitionLevels.add(definitionLevel);
+                repetitionLevels.add(repetitionLevel);
+            });
+        }
     }
 
-    private void skipValues(int valuesToRead)
+    private int skipValues(int valuesToRead)
     {
-        processValues(valuesToRead, this::skipValue);
+        return processValues(valuesToRead, this::skipValue);
     }
 
-    private void processValues(int valuesToRead, Runnable valueReader)
+    private int processValues(int valuesToRead, Runnable valueReader)
     {
         if (definitionLevel == EMPTY_LEVEL_VALUE && repetitionLevel == EMPTY_LEVEL_VALUE) {
             definitionLevel = definitionReader.readLevel();
             repetitionLevel = repetitionReader.readLevel();
         }
+        int rowCount = 0;
         int valueCount = 0;
         for (int i = 0; i < valuesToRead; i++) {
             do {
+                nextRow(repetitionLevel);
+                if (repetitionLevel == 0) {
+                    rowCount++;
+                }
                 valueReader.run();
                 valueCount++;
                 if (valueCount == remainingValueCountInPage) {
-                    updateValueCounts(valueCount);
+                    updateValueCounts(valueCount, 0);
+                    if (!readNextPage()) {
+                        return rowCount;
+                    }
+                    valueCount = 0;
+                }
+                repetitionLevel = repetitionReader.readLevel();
+                definitionLevel = definitionReader.readLevel();
+            }
+            while (repetitionLevel != 0);
+        }
+        updateValueCounts(valueCount, 0);
+        return rowCount;
+    }
+
+    /**
+     * When filtering using column indexes we might skip reading some pages for different columns. Because the rows are
+     * not aligned between the pages of the different columns it might be required to skip some values. The values (and the
+     * related rl and dl) are skipped based on the iterator of the required row indexes and the first row index of each
+     * page.
+     * For example:
+     *
+     * rows   col1   col2   col3
+     *      ┌──────┬──────┬──────┐
+     *   0  │  p0  │      │      │
+     *      ╞══════╡  p0  │  p0  │
+     *  20  │ p1(X)│------│------│
+     *      ╞══════╪══════╡      │
+     *  40  │ p2(X)│      │------│
+     *      ╞══════╡ p1(X)╞══════╡
+     *  60  │ p3(X)│      │------│
+     *      ╞══════╪══════╡      │
+     *  80  │  p4  │      │  p1  │
+     *      ╞══════╡  p2  │      │
+     * 100  │  p5  │      │      │
+     *      └──────┴──────┴──────┘
```

Comment:
This should be wrapped in `<pre> ... </pre>` so that the javadoc renders correctly

Reply:
Done